pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -0,0 +1,352 @@
+			for _, ep := range slice.Endpoints {
+				for _, a := range ep.Addresses {
+					if a == ip {
+						out = append(out, c.getEndpoints(podIP, ip, *port.Port, svcPort, svc))
+						// If the endpoint isn't ready, report this
+						if ep.Conditions.Ready != nil && !*ep.Conditions.Ready && c.metrics != nil {
+							c.metrics.AddMetric(model.ProxyStatusEndpointNotReady, proxy.ID, proxy, "")
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return out
+}
+
+func (e *endpointSliceController) InstancesByPort(c *Controller, svc *model.Service, reqSvcPort int,
+	labelsList labels.Collection) ([]*model.ServiceInstance, error) {
+	esLabelSelector := klabels.Set(map[string]string{discoveryv1alpha1.LabelServiceName: svc.Attributes.Name}).AsSelectorPreValidated()
+	slices, err := v1alpha1.NewEndpointSliceLister(e.cache.informer.GetIndexer()).EndpointSlices(svc.Attributes.Namespace).List(esLabelSelector)
+	if err != nil {
+		log.Infof("get endpoints(%s, %s) => error %v", svc.Attributes.Name, svc.Attributes.Namespace, err)
+		return nil, nil
+	}
+	if len(slices) == 0 {
+		return nil, nil
+	}
+
+	// Locate all ports in the actual service
+	svcPortEntry, exists := svc.Ports.GetByPort(reqSvcPort)
+	if !exists {
+		return nil, nil
+	}
+
+	var out []*model.ServiceInstance
+	for _, slice := range slices {
+		for _, e := range slice.Endpoints {
+			for _, a := range e.Addresses {
+				var podLabels labels.Instance
+				pod := c.pods.getPodByIP(a)
+				if pod != nil {
+					podLabels = configKube.ConvertLabels(pod.ObjectMeta)
+				}
+
+				// check that one of the input labels is a subset of the labels
+				if !labelsList.HasSubsetOf(podLabels) {
+					continue
+				}
+
+				sa, uid := "", ""
+				if pod != nil {
+					sa = kube.SecureNamingSAN(pod)
+					uid = createUID(pod.Name, pod.Namespace)
+				}
+				az := getLocalityFromTopology(e.Topology)
+				tlsMode := kube.PodTLSMode(pod)
+
+				// identify the port by name. K8S EndpointPort uses the service port name
+				for _, port := range slice.Ports {
+					var portNum uint32
+					if port.Port != nil {
+						portNum = uint32(*port.Port)
+					}
+
+					if port.Name == nil ||
+						svcPortEntry.Name == *port.Name {
+
+						out = append(out, &model.ServiceInstance{
+							Endpoint: model.NetworkEndpoint{
+								Address:     a,
+								Port:        int(portNum),
+								ServicePort: svcPortEntry,
+								UID:         uid,
+								Network:     c.endpointNetwork(a),
+								Locality:    az,
+							},
+							Service:        svc,
+							Labels:         podLabels,
+							ServiceAccount: sa,
+							TLSMode:        tlsMode,
+						})
+					}
+				}
+			}
+		}
+	}
+	return out, nil
+}
+
+func getLocalityFromTopology(topology map[string]string) string {
+	locality := topology[NodeRegionLabelGA]
+	if _, f := topology[NodeZoneLabelGA]; f {
+		locality += "/" + topology[NodeZoneLabelGA]
+	}
+	if _, f := topology[IstioSubzoneLabel]; f {
+		locality += "/" + topology[IstioSubzoneLabel]
+	}
+	return locality
```

Comment:
In istio we have `istio-locaity` label from pod should override the above, and also we have recently added `"topology.istio.io/subzone"` in node label. This makes it difficult to handle. 
It's better we can push this subzone label to k8s upstream. And make it in endpointslice's topology.

Reply:
@nmittler is working on seeing if we can get extra labels added upstream

Reply:
for now I can get the pod label though, as we still have to read the pod for service account and other things